Generating reports

dafuga opened the pull request:
Hey guys! 

A little PR that:

- Adds a method that returns random answers to each questions associated to a given review when that review is completed.

- Adds an action to generate a pdf report with a list of all answers by relationship_types.

Also, there seems to be an issue with the build on codeship. I think that the test is failing because there is no 'public/system/reports' folder on the codeship host. @sheldond would you know how to add it?

Thanks!!
Dan

sheldond commented:
Hey @dafuga I was able to take a look at this PR today, thanks!

The first few things we should address are:
- We should avoid writing raw SQL, is it possible to refactor the code so that there are no hardcoded SQL statements?
- We can't assume that a dir is going to exist in the environment that the app runs in, so if you want to write your file to a dir you'll need some way to handle creating the dir. You could automatically create the dir if it doesn't exist yet, or at minimum print out a friendly error message with instructions. 
- Any file written into the public dir will be publicly accessible (anyone on the internet can get it), so we should avoid putting files there, instead they could be written to a private dir owned by the app.

There a few things to consider when writing files to disk:
* If the files get really big then they'll need to be written to disk (as they are now), if they're not so big then maybe you can avoid writing to disk and just send them directly from memory (not ideal but maybe easier).
* If a file is written to disk then it will use up disk space and we will need an automated method of cleaning up files on disk to prevent the disk from filling up. If you go this route we could reuse something we've written before. 
* You may consider writing to a temporary file which is automatically removed after it's not needed anymore, Ruby core has Tempfile http://ruby-doc.org/stdlib-2.3.1/libdoc/tempfile/rdoc/Tempfile.html I'd recommend trying this approach first. :)

I haven't looked into Prawn much yet so I'm not sure how it behaves by default! Please let me know if you'd like any help with this!

Sheldon

dafuga commented:
Hey Sheldon,

Thanks for the feedback!

- I can try removing the sql statement and replace it with nested joins, but I was getting weird errors when I tried doing this, because FE uses an abstract class and that was causing me an error for some reason which is why I switched to a sql statement. I will give it a second shot though!

- For files, Josh wants us to use S3 for prod and staging and making these files publicly available is not an issue. For running tests though I will look into the Tempfile example that you provided.

Thanks :)
Daniel

sheldond commented:
Okay great thanks!

If the files will be uploaded to S3 then the tests can test that behaviour, and mocking or stubbing could be used to prevent real network requests.

Thanks!

dafuga commented:
Yes but actually for the case of reports, since we are generating a new file every time the report is downloaded. Wouldn't it be better to save the file to the hard drive and delete it right after the download?

sheldond commented:
Yes I think that makes sense!

dafuga commented:
I just looked into that and its a bit complicated because it take a few seconds often to download the file. I think that I will keep everything on S3 to keep things simple and use stubbing as you suggested.

sheldond commented:
Okay I think S3 is okay too. If it's helpful I can send an example using Tempfile with send_data.

dafuga commented:
Yeah if you could send an example that would be cool!! Thanks!

twinge commented:
In response to a comment I just noticed earlier in the thread, the
generated pdfs can't be public. They need to follow the same access rules
as the rest of the system. In other words, you should only be able to
download the pdf if you had access to click whatever link generated the pdf.

On Mon, May 9, 2016 at 6:53 PM, Daniel Fugere <notifications@github.com>
wrote:

> Yeah if you could send an example that would be cool!! Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/CruGlobal/global360/pull/17#issuecomment-218013619>
>

dafuga commented:
Ok so you would prefer using Tempfile for handling pdf generation? Or would you rather stick with S3 and add permissions?

sheldond commented:
@twinge got it! That is what I thought.

@dafuga if you decide to use Tempfile here is the example:

Model: https://github.com/ballistiq/downtown/blob/a1547b4d3ba4810f6fda2def85ad74d620df894a/app/models/concerns/csvable.rb
Controller: https://github.com/ballistiq/downtown/blob/a1547b4d3ba4810f6fda2def85ad74d620df894a/app/controllers/admin/variant_serials_controller.rb

twinge commented:
Tempfile works fine if what's being downloaded is a generated file. If it's
a file that's going to be downloaded several times, s3 becomes attractive
for performance reasons.

On Tue, May 10, 2016 at 11:04 AM, Sheldon <notifications@github.com> wrote:

> @twinge <https://github.com/twinge> got it! That is what I thought.
>
> @dafuga <https://github.com/dafuga> if you decide to use Tempfile here is
> the example:
>
> Model:
> https://github.com/ballistiq/downtown/blob/a1547b4d3ba4810f6fda2def85ad74d620df894a/app/models/concerns/csvable.rb
> Controller:
> https://github.com/ballistiq/downtown/blob/a1547b4d3ba4810f6fda2def85ad74d620df894a/app/controllers/admin/variant_serials_controller.rb
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/CruGlobal/global360/pull/17#issuecomment-218186450>
>

dafuga commented:
Ok sounds good, thanks!

dafuga commented:
Hey @sheldond , I just spent another hour trying to change my raw sql statement into a nested joins with conditionals, however joins really don't seem to work nicely with an abstract class and since the Fe::AnswerSheet model is an abstract class, it won't work in our case. I don't like writing raw SQL statements, but sometimes I think that it's the best solution.

sheldond commented:
@dafuga okay I understand and thanks for trying! We may consider abstracting these Fe queries into objects at some point.

dafuga commented:
Not a bad idea, I will keep that in mind. Thanks!

dafuga commented:
Ok @sheldond so I switched to a Temp file for report generation and I also added S3 upload for the production environment and everything seems to be working fine.
